# TODO 1 (solved)

Wow, it looks like if the public `srtp_protect()` API and so on now requires extra arguments for the target. Note that in `SrtpSession.cpp` we do this:

```c++
std::memcpy(encryptBuffer, *data, *len);

const srtp_err_status_t err = srtp_protect(this->session, encryptBuffer, len);
```

But now it requires 6 args instead:

```
../../../src/RTC/SrtpSession.cpp:228:33: error: no matching function for call to 'srtp_protect'
  228 |                 const srtp_err_status_t err = srtp_protect(this->session, encryptBuffer, len);
      |                                               ^~~~~~~~~~~~
../../../subprojects/libsrtp-3.0.0-beta/include/srtp.h:433:19: note: candidate function not viable: requires 6 arguments, but 3 were provided
  433 | srtp_err_status_t srtp_protect(srtp_t ctx,
      |                   ^            ~~~~~~~~~~~
  434 |                                const uint8_t *rtp,
      |                                ~~~~~~~~~~~~~~~~~~~
  435 |                                size_t rtp_len,
      |                                ~~~~~~~~~~~~~~~
  436 |                                uint8_t *srtp,
      |                                ~~~~~~~~~~~~~~
  437 |                                size_t *srtp_len,
      |                                ~~~~~~~~~~~~~~~~~
  438 |                                size_t mki_index);
      |                                ~~~~~~~~~~~~~~~~
../../../src/RTC/SrtpSession.cpp:247:33: error: no matching function for call to 'srtp_unprotect'
  247 |                 const srtp_err_status_t err = srtp_unprotect(this->session, data, len);
      |                                               ^~~~~~~~~~~~~~
../../../subprojects/libsrtp-3.0.0-beta/include/srtp.h:485:19: note: candidate function not viable: requires 5 arguments, but 3 were provided
  485 | srtp_err_status_t srtp_unprotect(srtp_t ctx,
      |                   ^              ~~~~~~~~~~~
  486 |                                  const uint8_t *srtp,
      |                                  ~~~~~~~~~~~~~~~~~~~~
  487 |                                  size_t srtp_len,
      |                                  ~~~~~~~~~~~~~~~~
  488 |                                  uint8_t *rtp,
      |                                  ~~~~~~~~~~~~~
  489 |                                  size_t *rtp_len);
      |                                  ~~~~~~~~~~~~~~~
../../../src/RTC/SrtpSession.cpp:273:33: error: no matching function for call to 'srtp_protect_rtcp'
  273 |                 const srtp_err_status_t err = srtp_protect_rtcp(this->session, EncryptBuffer, len);
      |                                               ^~~~~~~~~~~~~~~~~
../../../subprojects/libsrtp-3.0.0-beta/include/srtp.h:1162:19: note: candidate function not viable: requires 6 arguments, but 3 were provided
 1162 | srtp_err_status_t srtp_protect_rtcp(srtp_t ctx,
      |                   ^                 ~~~~~~~~~~~
 1163 |                                     const uint8_t *rtcp,
      |                                     ~~~~~~~~~~~~~~~~~~~~
 1164 |                                     size_t rtcp_len,
      |                                     ~~~~~~~~~~~~~~~~
 1165 |                                     uint8_t *srtcp,
      |                                     ~~~~~~~~~~~~~~~
 1166 |                                     size_t *srtcp_len,
      |                                     ~~~~~~~~~~~~~~~~~~
 1167 |                                     size_t mki_index);
      |                                     ~~~~~~~~~~~~~~~~
../../../src/RTC/SrtpSession.cpp:292:33: error: no matching function for call to 'srtp_unprotect_rtcp'
  292 |                 const srtp_err_status_t err = srtp_unprotect_rtcp(this->session, data, len);
      |                                               ^~~~~~~~~~~~~~~~~~~
../../../subprojects/libsrtp-3.0.0-beta/include/srtp.h:1212:19: note: candidate function not viable: requires 5 arguments, but 3 were provided
 1212 | srtp_err_status_t srtp_unprotect_rtcp(srtp_t ctx,
      |                   ^                   ~~~~~~~~~~~
 1213 |                                       const uint8_t *srtcp,
      |                                       ~~~~~~~~~~~~~~~~~~~~~
 1214 |                                       size_t srtcp_len,
      |                                       ~~~~~~~~~~~~~~~~~
 1215 |                                       uint8_t *rtcp,
      |                                       ~~~~~~~~~~~~~~
 1216 |                                       size_t *rtcp_len);
      |                                       ~~~~~~~~~~~~~~~~
```

Signature of `srtp_protect()`:

```c++
/**
 * @brief srtp_protect() is the Secure RTP sender-side packet processing
 * function.
 *
 * The function call srtp_protect(ctx, rtp, rtp_len, srtp, srtp_len, mki_index)
 * applies SRTP protection to the RTP packet rtp (which has length rtp_len)
 * using the SRTP context ctx.  If srtp_err_status_ok is returned, then srtp
 * points to the resulting SRTP packet and *srtp_len is the number of
 * octets in that packet; otherwise, no assumptions should be made
 * about the value of either data elements.
 *
 * The sequence numbers of the RTP packets presented to this function
 * need not be consecutive, but they @b must be out of order by less
 * than 2^15 = 32,768 packets.
 *
 * @warning This function assumes that the RTP packet is aligned on a 32-bit
 * boundary.
 *
 * @param ctx is the SRTP context to use in processing the packet.
 *
 * @param rtp is a pointer to the RTP packet.
 *
 * @param rtp_len is the length in octets of the complete RTP
 * packet (header and body).
 *
 * @param srtp is a pointer to a buffer that after the function returns will
 * contain the complete SRTP packet. The value of srtp can be the same as rtp to
 * support in-place io.
 *
 * @param srtp_len is a pointer to the length in octets of the srtp buffer
 * before the function call, and of the complete SRTP packet after the call, if
 * srtp_err_status_ok was returned. Otherwise, the value of the data to which it
 * points is undefined.
 *
 * @param mki_index integer value specifying which set of session keys should be
 * used if use_mki in the policy was set to true. Otherwise ignored.
 *
 * @return
 *    - srtp_err_status_ok            no problems
 *    - srtp_err_status_replay_fail   rtp sequence number was non-increasing
 *    - srtp_err_status_buffer_small  the srtp buffer is too small for the SRTP
 * packet
 *    - @e other                 failure in cryptographic mechanisms
 */
srtp_err_status_t srtp_protect(srtp_t ctx,
                               const uint8_t *rtp,
                               size_t rtp_len,
                               uint8_t *srtp,
                               size_t *srtp_len,
                               size_t mki_index);
```

--- a/worker/src/RTC/SrtpSession.cpp
+++ b/worker/src/RTC/SrtpSession.cpp
@@ -8,7 +8,7 @@
 #endif
 #include "Logger.hpp"
 #include "MediaSoupErrors.hpp"
-#include <cstring> // std::memset(), std::memcpy()
+#include <cstring> // std::memset()
 
 namespace RTC
 {
@@ -202,6 +202,7 @@ namespace RTC
 		}
 
 		uint8_t* encryptBuffer = EncryptBuffer;
+		size_t encryptLen      = EncryptBufferSize;
 
 #ifdef MS_LIBURING_SUPPORTED
 		if (DepLibUring::IsEnabled())
@@ -217,15 +218,20 @@ namespace RTC
 			if (sendBuffer)
 			{
 				encryptBuffer = sendBuffer;
+				encryptLen    = DepLibUring::SendBufferSize;
 			}
 		}
 
 	protect:
 #endif
 
-		std::memcpy(encryptBuffer, *data, *len);
-
-		const srtp_err_status_t err = srtp_protect(this->session, encryptBuffer, len);
+		const srtp_err_status_t err = srtp_protect(
+		  /*srtp_t ctx*/ this->session,
+		  /*const uint8_t* rtp*/ *data,
+		  /*size_t rtp_len*/ *len,
+		  /*uint8_t* srtp*/ encryptBuffer,
+		  /*size_t* srtp_len*/ std::addressof(encryptLen),
+		  /*size_t mki_index*/ 0);
 
 		if (DepLibSRTP::IsError(err))
 		{
@@ -234,8 +240,9 @@ namespace RTC
 			return false;
 		}
 
-		// Update the given data pointer.
+		// Update the given data pointer and len.
 		*data = const_cast<const uint8_t*>(encryptBuffer);
+		*len  = encryptLen;
 
 		return true;
 	}
@@ -244,7 +251,14 @@ namespace RTC
 	{
 		MS_TRACE();
 
-		const srtp_err_status_t err = srtp_unprotect(this->session, data, len);
+		size_t decryptLen = *len;
+
+		const srtp_err_status_t err = srtp_unprotect(
+		  /*srtp_t ctx*/ this->session,
+		  /*const uint8_t* srtp*/ data,
+		  /*size_t srtp_len*/ *len,
+		  /*uint8_t* rtp*/ data,
+		  /*size_t* rtp_len*/ std::addressof(decryptLen));
 
 		if (DepLibSRTP::IsError(err))
 		{
@@ -252,6 +266,9 @@ namespace RTC
 
 			return false;
 		}
+
+		// Update the given len.
+		*len = decryptLen;
 
 		return true;
 	}
@@ -268,9 +285,16 @@ namespace RTC
 			return false;
 		}
 
-		std::memcpy(EncryptBuffer, *data, *len);
+		uint8_t* encryptBuffer = EncryptBuffer;
+		size_t encryptLen      = EncryptBufferSize;
 
-		const srtp_err_status_t err = srtp_protect_rtcp(this->session, EncryptBuffer, len);
+		const srtp_err_status_t err = srtp_protect_rtcp(
+		  /*srtp_t ctx*/ this->session,
+		  /*const uint8_t* rtcp*/ *data,
+		  /*size_t rtcp_len*/ *len,
+		  /*uint8_t* srtcp*/ encryptBuffer,
+		  /*size_t* srtcp_len*/ std::addressof(encryptLen),
+		  /*size_t mki_index*/ 0);
 
 		if (DepLibSRTP::IsError(err))
 		{
@@ -279,8 +303,9 @@ namespace RTC
 			return false;
 		}
 
-		// Update the given data pointer.
+		// Update the given data pointer and len.
 		*data = (const uint8_t*)EncryptBuffer;
+		*len  = encryptLen;
 
 		return true;
 	}
@@ -289,7 +314,14 @@ namespace RTC
 	{
 		MS_TRACE();
 
-		const srtp_err_status_t err = srtp_unprotect_rtcp(this->session, data, len);
+		size_t decryptLen = *len;
+
+		const srtp_err_status_t err = srtp_unprotect_rtcp(
+		  /*srtp_t ctx*/ this->session,
+		  /*const uint8_t* srtcp*/ data,
+		  /*size_t srtcp_len*/ *len,
+		  /*uint8_t* rtcp*/ data,
+		  /*size_t* rtcp_len*/ std::addressof(decryptLen));
 
 		if (DepLibSRTP::IsError(err))
 		{
@@ -297,6 +329,9 @@ namespace RTC
 
 			return false;
 		}
+
+		// Update the given len.
+		*len = decryptLen;
 
 		return true;
 	}

